### ISSUE TITLE
less(1) fails with a CP2 exception

### ISSUE THREAD

**Issue by rwatson** [CTSRD-CHERI/cheribsd#161]
`less(1)` when compiled with CheriABI:

```
diff --git a/usr.bin/less/Makefile b/usr.bin/less/Makefile
index b6adf2c..b69549e 100644
--- a/usr.bin/less/Makefile
+++ b/usr.bin/less/Makefile
@@ -1,5 +1,9 @@
 # $FreeBSD$

+WANT_CHERI=pure
+WANT_DUMP=yes
+NO_SHARED=yes
+
 PROG=  less
 SRCS=  main.c screen.c brac.c ch.c charset.c cmdbuf.c command.c cvt.c \
        decode.c edit.c filename.c forwback.c help.c ifile.c input.c \
```

Crashes on Qemu with:

```
root@qemu-cheri:~ # less
CHERI cause: ExcCode: 0x01 RegNum: $c03 (length violation)
$c00: v:1 s:0 p:7fff807d b:0000000000000000 l:0000010000000000 o:0 t:0
$c01: v:1 s:0 p:7fff807d b:00000001201b6240 l:0000000000000004 o:0 t:0
$c02: v:1 s:0 p:7fff807d b:00000001201c1400 l:0000000000000008 o:0 t:0
$c03: v:1 s:0 p:7fff807d b:000000000041e010 l:0000000000000008 o:0 t:0
$c04: v:1 s:0 p:7fff807d b:00000001201c1408 l:0000000000000ea0 o:0 t:0
$c05: v:1 s:0 p:7fff807d b:0000007fffffd520 l:0000000000000020 o:0 t:0
$c06: v:1 s:0 p:00008055 b:0000007fffffd040 l:0000000000000040 o:0 t:0
$c07: v:1 s:0 p:7fff807d b:0000007fffff3ca0 l:0000000000008000 o:0 t:0
$c08: v:1 s:0 p:7fff807d b:00000001201baf80 l:0000000000000004 o:0 t:0
$c09: v:1 s:0 p:7fff807d b:00000001201bafa0 l:0000000000000020 o:0 t:0
$c10: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c11: v:1 s:0 p:7fff807d b:0000007fff7ff000 l:00000000007ff3c0 o:0 t:0
$c12: v:1 s:0 p:7fff8017 b:0000000000000000 l:0000010000000000 o:1200d8170 t:0
$c13: v:1 s:0 p:00008055 b:0000007fffffd720 l:0000000000000010 o:0 t:0
$c14: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c15: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c16: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c17: v:1 s:0 p:7fff8017 b:0000000000000000 l:0000010000000000 o:12003918c t:0
$c18: v:1 s:0 p:7fff807d b:00000001201b6220 l:0000000000000020 o:0 t:0
$c19: v:1 s:0 p:7fff807d b:00000001201aeda0 l:0000000000000001 o:0 t:0
$c20: v:1 s:0 p:7fff807d b:00000001201bfd40 l:0000000000000020 o:0 t:0
$c21: v:1 s:0 p:7fff807d b:0000007fffffe460 l:0000000000000040 o:20 t:0
$c22: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c23: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c24: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c26: v:1 s:0 p:7fff807d b:0000000000000000 l:0000010000000000 o:0 t:0
$c31: v:1 s:0 p:7fff8017 b:0000000000000000 l:0000010000000000 o:1200391f0 t:0
Jul 19 11:03:54 qemu-cheri kernel: USER_CHERI_EXCEPTION: pid 563 tid 100033 (less), uid 0: CP2 fault (type 0x32)
Jul 19 11:03:54 qemu-cheri kernel: Trapframe Register Dump:
Jul 19 11:03:54 qemu-cheri kernel: zero: 0      at: 0x8 v0: 0x18        v1: 0x2
Jul 19 11:03:54 qemu-cheri kernel: a0: 0xffffffffffffffff       a1: 0   a2: 0x8 a3: 0
Jul 19 11:03:54 qemu-cheri kernel: a4: 0x4      a5: 0x8 a6: 0   a6: 0xffffffffffffffff
Jul 19 11:03:54 qemu-cheri kernel: t0: 0x2      t1: 0   t2: 0xffffffffffffffff  t3: 0x5555555555555555
Jul 19 11:03:54 qemu-cheri kernel: t8: 0x8      t9: 0x12010be18 s0: 0   s1: 0xffffffffffffffff
Jul 19 11:03:54 qemu-cheri kernel: s2: 0        s3: 0   s4: 0   s5: 0
Jul 19 11:03:54 qemu-cheri kernel: s6: 0        s7: 0   k0: 0   k1: 0
Jul 19 11:03:54 qemu-cheri kernel: gp: 0x1201af110      sp: 0x7fe720    s8: 0x7fe720    ra: 0
Jul 19 11:03:54 qemu-cheri kernel: sr: 0x408084b3       mullo: 0x10     mulhi: 0        badvaddr: 0x1200391f0
Jul 19 11:03:54 qemu-cheri kernel: cause: 0x48  pc: 0x1200391f0
Signal 34 (core dumped)
```

Interestingly, it crashes slightly differently on a recent CHERI bitfile:

```
> ./less
cpuid = 0
USER_CHERI_EXCEPTION: pid 30514 tid 100050 (less), uid 999: CP2 fault (type 0x32)
Trapframe Register Dump:
    zero: 0 at: 0x1200c83c0 v0: 0   v1: 0x64
    a0: 0xe a1: 0x52    a2: 0x4c    a3: 0
    a4: 0x2 a5: 0x8 a6: 0   a6: 0xffffffffffffffff
    t0: 0x1 t1: 0   t2: 0xffffffffffffffff  t3: 0x5555555555555555
    t8: 0x3333333333333333  t9: 0x120005920 s0: 0x1 s1: 0x20
    s2: 0   s3: 0   s4: 0   s5: 0
    s6: 0   s7: 0   k0: 0   k1: 0
    gp: 0x1201af110 sp: 0x7fed60    s8: 0x7fed60    ra: 0
    sr: 0x4000fcf3  mullo: 0x60 mulhi: 0    badvaddr: 0x12001ead8
    cause: 0x148    pc: 0x1200c83c4
cpuid = 0
CHERI cause: ExcCode: 0x02 RegNum: $c03 (tag violation)
$c00: v:1 s:0 p:7fff807d b:0000000000000000 l:0000010000000000 o:0 t:0
$c01: v:1 s:0 p:7fff807d b:0000007fffffde20 l:0000000000000020 o:0 t:0
$c02: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c03: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c04: v:1 s:0 p:7fff807d b:000000012013360e l:000000000000000f o:0 t:0
$c05: v:1 s:0 p:7fff807d b:0000007fffffde80 l:0000000000000020 o:0 t:0
$c06: v:1 s:0 p:7fff807d b:00000001201aefc0 l:0000000000000020 o:0 t:0
$c07: v:1 s:0 p:7fff807d b:0000007fffff5760 l:0000000000008000 o:0 t:0
$c08: v:1 s:0 p:7fff807d b:000000012013908f l:0000000000000001 o:0 t:0
$c09: v:1 s:0 p:7fff807d b:0000007fffffdf00 l:0000000000000020 o:0 t:0
$c10: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c11: v:1 s:0 p:7fff807d b:0000007fff7ff000 l:00000000007ff2e0 o:0 t:0
$c12: v:1 s:0 p:7fff8017 b:0000000000000000 l:0000010000000000 o:1200c83c0 t:0
$c13: v:1 s:0 p:00008055 b:0000007fffffde20 l:0000000000000020 o:0 t:0
$c14: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c15: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c16: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c17: v:1 s:0 p:7fff8017 b:0000000000000000 l:0000010000000000 o:120005964 t:0
$c18: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c19: v:1 s:0 p:7fff807d b:000000012013360e l:000000000000000f o:0 t:0
$c20: v:1 s:0 p:7fff807d b:0000007fffffde80 l:0000000000000020 o:0 t:0
$c21: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c22: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c23: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c24: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c26: v:1 s:0 p:7fff807d b:0000000000000000 l:0000010000000000 o:0 t:0
$c31: v:1 s:0 p:7fff8017 b:0000000000000000 l:0000010000000000 o:1200c83c4 t:0
pid 30514 (less), uid 999: exited on signal 34
Signal 34
```

Pursing the Qemu-CHERI trace, the problem arises in `less(1)`'s `pos_init()` function:

```
0000000120039020 <pos_init>:
/*
 * Allocate or reallocate the position table.
 */
        public void
pos_init()
{
   120039020:   67bdff80        daddiu  sp,sp,-128
   120039024:   ebcbe87b        csd     s8,sp,120(c11)
   120039028:   eb8be873        csd     gp,sp,112(c11)
   12003902c:   ea2be86b        csd     s1,sp,104(c11)
   120039030:   ea0be863        csd     s0,sp,96(c11)
   120039034:   fa4be804        csc     c18,sp,64(c11)
   120039038:   fa2be802        csc     c17,sp,32(c11)
   12003903c:   03a0f025        move    s8,sp
   120039040:   49b96002        cgetoffset      t9,c12
   120039044:   3c010017        lui     at,0x17
   120039048:   0039082d        daddu   at,at,t9
   12003904c:   643c60f0        daddiu  gp,at,24816
        struct scrpos scrpos;

        if (sc_height <= table_size)
   120039050:   3c010000        lui     at,0x0
   120039054:   003c082d        daddu   at,at,gp
   120039058:   dc21bce8        ld      at,-17176(at)
   12003905c:   dc210000        ld      at,0(at)
   120039060:   3c020000        lui     v0,0x0
   120039064:   005c102d        daddu   v0,v0,gp
   120039068:   dc42ecb8        ld      v0,-4936(v0)
   12003906c:   3c030000        lui     v1,0x0
   120039070:   007c182d        daddu   v1,v1,gp
   120039074:   dc638e48        ld      v1,-29112(v1)
   120039078:   48810087        cfromptr        c1,c0,v0
   12003907c:   48210840        csetbounds      c1,c1,at
   120039080:   c8410006        clw     v0,zero,0(c1)
   120039084:   488100c7        cfromptr        c1,c0,v1
   120039088:   64010004        daddiu  at,zero,4
   12003908c:   48210840        csetbounds      c1,c1,at
   120039090:   c8210006        clw     at,zero,0(c1)
   120039094:   0022082a        slt     at,at,v0
   120039098:   10200060        beqz    at,12003921c <pos_init+0x1fc>
   12003909c:   00000000        nop
                return;
        /*
         * If we already have a table, remember the first line in it
         * before we free it, so we can copy that line to the new table.
         */
        if (table != NULL)
   1200390a0:   3c010000        lui     at,0x0
   1200390a4:   003c082d        daddu   at,at,gp
   1200390a8:   dc218e40        ld      at,-29120(at)
   1200390ac:   48810047        cfromptr        c1,c0,at
   1200390b0:   64010020        daddiu  at,zero,32
   1200390b4:   48320840        csetbounds      c18,c1,at
   1200390b8:   d8720000        clc     c3,zero,0(c18)
   1200390bc:   48810007        cfromptr        c1,c0,zero
   1200390c0:   49c11840        ceq     at,c3,c1
   1200390c4:   1420000d        bnez    at,1200390fc <pos_init+0xdc>
   1200390c8:   00000000        nop


        /*
         * Find the first line on the screen which has something on it,
         * and return the screen line number and the file position.
         */
        for (i = 0; i < sc_height;  i++)
   1200390cc:   00021000        sll     v0,v0,0x0
                if (table[i] != NULL_POSITION)
   1200390d0:   6403ffff        daddiu  v1,zero,-1
   1200390d4:   64100000        daddiu  s0,zero,0

        /*
         * Find the first line on the screen which has something on it,
         * and return the screen line number and the file position.
         */
        for (i = 0; i < sc_height;  i++)
   1200390d8:   0202082a        slt     at,s0,v0
   1200390dc:   1020000a        beqz    at,120039108 <pos_init+0xe8>
   1200390e0:   00000000        nop
                if (table[i] != NULL_POSITION)
   1200390e4:   001008f8        dsll    at,s0,0x3
                {
                        scrpos->ln = i+1;
                        scrpos->pos = table[i];
   1200390e8:   ca230803        cld     s1,at,0(c3)
   1200390ec:   1223fffa        beq     s1,v1,1200390d8 <pos_init+0xb8>
   1200390f0:   66100001        daddiu  s0,s0,1
   1200390f4:   0800e444        j       120039110 <pos_init+0xf0>
   1200390f8:   00000000        nop
   1200390fc:   64100000        daddiu  s0,zero,0
   120039100:   0800e454        j       120039150 <pos_init+0x130>
   120039104:   6411ffff        daddiu  s1,zero,-1
   120039108:   64100000        daddiu  s0,zero,0
        /*
         * Find the first line on the screen which has something on it,
         * and return the screen line number and the file position.
         */
        for (i = 0; i < sc_height;  i++)
                if (table[i] != NULL_POSITION)
   12003910c:   6411ffff        daddiu  s1,zero,-1
         * before we free it, so we can copy that line to the new table.
         */
        if (table != NULL)
        {
                get_scrpos(&scrpos);
                free((char*)table);
   120039110:   3c010000        lui     at,0x0
   120039114:   003c082d        daddu   at,at,gp
   120039118:   dc21bae8        ld      at,-17688(at)
   12003911c:   480c09ff        cgetpccsetoffset        c12,at
   120039120:   48f16000        cjalr   c12,c17
   120039124:   00000000        nop
        } else
                scrpos.pos = NULL_POSITION;
        table = (POSITION *) ecalloc(sc_height, sizeof(POSITION));
   120039128:   3c010000        lui     at,0x0
   12003912c:   003c082d        daddu   at,at,gp
   120039130:   dc21bce8        ld      at,-17176(at)
   120039134:   3c020000        lui     v0,0x0
   120039138:   005c102d        daddu   v0,v0,gp
   12003913c:   dc42ecb8        ld      v0,-4936(v0)
   120039140:   dc210000        ld      at,0(at)
   120039144:   48810087        cfromptr        c1,c0,v0
   120039148:   48210840        csetbounds      c1,c1,at
   12003914c:   c8410006        clw     v0,zero,0(c1)
   120039150:   00020800        sll     at,v0,0x0
   120039154:   49a15f40        cincoffset      c1,c11,sp
   120039158:   e82be803        csd     at,sp,0(c11)
   12003915c:   64010010        daddiu  at,zero,16
   120039160:   48210840        csetbounds      c1,c1,at
   120039164:   3c010000        lui     at,0x0
   120039168:   003c082d        daddu   at,at,gp
   12003916c:   dc21f158        ld      at,-3752(at)
   120039170:   3402ffd7        li      v0,0xffd7
   120039174:   488d0880        candperm        c13,c1,v0
   120039178:   67a20008        daddiu  v0,sp,8
   12003917c:   480c09ff        cgetpccsetoffset        c12,at
   120039180:   64010008        daddiu  at,zero,8
   120039184:   48f16000        cjalr   c12,c17
   120039188:   e82b1003        csd     at,v0,0(c11)
        table_size = sc_height;
   12003918c:   3c010000        lui     at,0x0
   120039190:   003c082d        daddu   at,at,gp
   120039194:   dc21bce8        ld      at,-17176(at)
   120039198:   3c020000        lui     v0,0x0
   12003919c:   005c102d        daddu   v0,v0,gp
   1200391a0:   3c030000        lui     v1,0x0
   1200391a4:   007c182d        daddu   v1,v1,gp
   1200391a8:   dc42ecb8        ld      v0,-4936(v0)
   1200391ac:   dc210000        ld      at,0(at)
        {
                get_scrpos(&scrpos);
                free((char*)table);
        } else
                scrpos.pos = NULL_POSITION;
        table = (POSITION *) ecalloc(sc_height, sizeof(POSITION));
   1200391b0:   f8720000        csc     c3,zero,0(c18)
        table_size = sc_height;
   1200391b4:   dc638e48        ld      v1,-29112(v1)
   1200391b8:   48810087        cfromptr        c1,c0,v0
   1200391bc:   48210840        csetbounds      c1,c1,at
   1200391c0:   c8410006        clw     v0,zero,0(c1)
   1200391c4:   488100c7        cfromptr        c1,c0,v1
   1200391c8:   64010004        daddiu  at,zero,4
   1200391cc:   48210840        csetbounds      c1,c1,at
        public void
pos_clear()
{
        register int i;

        for (i = 0;  i < sc_height;  i++)
   1200391d0:   18400008        blez    v0,1200391f4 <pos_init+0x1d4>
   1200391d4:   e8410002        csw     v0,zero,0(c1)
   1200391d8:   00021000        sll     v0,v0,0x0
   1200391dc:   64030000        daddiu  v1,zero,0
                table[i] = NULL_POSITION;
   1200391e0:   6404ffff        daddiu  a0,zero,-1
   1200391e4:   000308f8        dsll    at,v1,0x3
        public void
pos_clear()
{
        register int i;

        for (i = 0;  i < sc_height;  i++)
   1200391e8:   64630001        daddiu  v1,v1,1
   1200391ec:   1443fffd        bne     v0,v1,1200391e4 <pos_init+0x1c4>
   1200391f0:   e8830803        csd     a0,at,0(c3)
        } else
                scrpos.pos = NULL_POSITION;
        table = (POSITION *) ecalloc(sc_height, sizeof(POSITION));
        table_size = sc_height;
        pos_clear();
        if (scrpos.pos != NULL_POSITION)
   1200391f4:   6401ffff        daddiu  at,zero,-1
   1200391f8:   12210008        beq     s1,at,12003921c <pos_init+0x1fc>
   1200391fc:   00000000        nop
                table[scrpos.ln-1] = scrpos.pos;
   120039200:   0010083c        dsll32  at,s0,0x0
   120039204:   3c02ffff        lui     v0,0xffff
   120039208:   3442ffff        ori     v0,v0,0xffff
   12003920c:   0002103c        dsll32  v0,v0,0x0
   120039210:   0022082d        daddu   at,at,v0
   120039214:   00010f7b        dsra    at,at,0x1d
   120039218:   ea230803        csd     s1,at,0(c3)
}
   12003921c:   03c0e825        move    sp,s8
   120039220:   da2be802        clc     c17,sp,32(c11)
   120039224:   da4be804        clc     c18,sp,64(c11)
   120039228:   ca0be863        cld     s0,sp,96(c11)
   12003922c:   ca2be86b        cld     s1,sp,104(c11)
   120039230:   cb8be873        cld     gp,sp,112(c11)
   120039234:   cbcbe87b        cld     s8,sp,120(c11)
   120039238:   49008800        cjr     c17
   12003923c:   67bd0080        daddiu  sp,sp,128
```

From the trace:

```
#
# In function pos_init():
#
# Exception due to write beyond the end of $c3's bounds:
#
# at = 0000000000000008
# a0 = ffffffffffffffff
#
# c3 = |v:1 s:0 p:7fff807d b:000000000041e010 l:0000000000000008 |o:0000000000000000 t:0
#
0x00000001200391f0:  csd        a0,at,0(c3)
mips_cpu_do_interrupt enter: PC 00000001200391f0 EPC 0000000120005a40 precise coprocessor 2 exception
    Write BadVAddr = 00000001200391f0
```

This appears to occur because the `less(1)` header declaring function prototypes, `funcs.h`, does not declare arguments to the functions. I am told that this can work if variadic and non-variadic calling conventions are the same for an ABI .. which in CheriABI they are not. As a result, the `ecalloc` function does not receive argument registers properly describing the count and size of proposed allocations, causing it to use .. other .. values, such as 8 above.

It is unclear why Qemu-CHERI and the bitfile are failing differently, and this likely deserves further investigation.

**Comment by rwatson** [comment]
Huh. 298 K&R function prototypes in ```funcs.h```. Correcting this problem with ```ecalloc()``` immediately leads to a near-identical problem with ```save()```. It may be that they all need type information before ```less(1)``` will work in CheriABI.

**Comment by rwatson** [comment]
@brooksdavis resolved this problem in 0ae56560b38b444bf023e4cb4c6eeffa186968e4 by adding the missing argument types to ```funcs.h```.
